### PR TITLE
Visualizations: Remove singlestat panel

### PR DESCRIPTION
**Pull request by dprokop**
- [x] Removes singlestat code
- [x] Automatically migrates singlestat -> stat panel
- [x] Automatically migrates singlestat -> gauge panel

Wonder if we should have some kind of notifications for the user to let them know what has happened with their panel that was automatically migrated?

### CODE REVIEW

File: public/app/features/plugins/plugin_loader.ts [hunk 1]
```diff
@@ -222,6 +227,10 @@ interface PanelCache {
 const panelCache: PanelCache = {};
 
 export function importPanelPlugin(id: string): Promise<grafanaData.PanelPlugin> {
+  if (DEPRECATED_PANELS[id]) {
+    id = DEPRECATED_PANELS[id];
+  }
```
Comment (torkelo) [inline public/app/features/plugins/plugin_loader.ts:232]: Is this all we need? won't we need to change the panel type property in the dashboard migrator

Comment (dprokop) [reply public/app/features/plugins/plugin_loader.ts:232]: Proobably, thought it's handled by the panel migrator (setMigrationHandler), but it's not

Comment (dprokop) [reply public/app/features/plugins/plugin_loader.ts:232]: @torkelo yup -> you are right:
image

**Comment by Clarity-89** [comment]
Related issue: https://github.com/grafana/grafana/issues/25527.

File: public/app/features/plugins/plugin_loader.ts (hunk 1 above)
Comment (dprokop) [reply public/app/features/plugins/plugin_loader.ts:232]: definitely will require more changes!

File: public/app/features/dashboard/state/actions.ts [hunk 2]
```diff
@@ -119,10 +121,22 @@ export function removeDashboard(uri: string): ThunkResult<void> {
...
+    let plugin = getStore().plugins.panels[pluginToLoad];
 
     if (!plugin) {
-      plugin = await dispatch(loadPanelPlugin(panel.type));
+      plugin = await dispatch(loadPanelPlugin(pluginToLoad));
+    }
+
+    if (isDeprecated) {
+      dispatch(changePanelPlugin(panel, DEPRECATED_PANELS[panel.type]));
     }
```
Comment (torkelo) [inline public/app/features/dashboard/state/actions.ts:140]: think this should be done in DashboardMigrator

File: public/app/features/dashboard/state/actions.ts [hunk 3]
```diff
@@ -119,10 +121,22 @@ export function removeDashboard(uri: string): ThunkResult<void> {
...
+
+    let plugin = getStore().plugins.panels[pluginToLoad];
 
     if (!plugin) {
-      plugin = await dispatch(loadPanelPlugin(panel.type));
+      plugin = await dispatch(loadPanelPlugin(pluginToLoad));
+    }
+
+    if (isDeprecated) {
+      dispatch(changePanelPlugin(panel, pluginToLoad));
```
Comment (torkelo) [inline public/app/features/dashboard/state/actions.ts:139]: Not sure I understand this condition, you change plugin but you already switched pluginToLoad to the new panel

Comment (torkelo) [reply public/app/features/dashboard/state/actions.ts:139]: oh wait now I undersand

Comment (torkelo) [reply public/app/features/dashboard/state/actions.ts:139]: think you need to add an await here as changePanelPlugin is async

**Review by torkelo — APPROVED** [review]
Looks good, added 1 nits, and 1 potential source of bugs

File: public/app/features/dashboard/state/actions.ts [hunk 4]
```diff
@@ -119,10 +121,22 @@ export function removeDashboard(uri: string): ThunkResult<void> {
...
-    let plugin = getStore().plugins.panels[panel.type];
+    let pluginToLoad = panel.type;
+
+    const isDeprecated = isDeprecatedPanel(panel.type);
+
+    if (isDeprecated) {
+      pluginToLoad = DEPRECATED_PANELS[panel.type](panel);
+    }
+
+    let plugin = getStore().plugins.panels[pluginToLoad];
```
[the remark this answers is not in the document]
Comment (dprokop) [reply public/app/features/dashboard/state/actions.ts:132]: ah yeah, this sounds like a good idea

File: public/app/core/constants.ts [hunk 5]
```diff
@@ -12,3 +15,18 @@ export const LS_PANEL_COPY_KEY = 'panel-copy';
 export const PANEL_BORDER = 2;
 
 export const EDIT_PANEL_ID = 23763571993;
+
+export const DEPRECATED_PANELS: Record<string, (panel: PanelModel) => string> = {
+  singlestat: (panel: PanelModel) => {
+    // If 'grafana-singlestat-panel' exists, move to that
```
Comment (ryantxu) [inline public/app/core/constants.ts:21]: See: https://github.com/grafana/singlestat-panel

For some reason, this does not seem to keep all the settings, but it does migrate :man_shrugging:

Comment (ryantxu) [reply public/app/core/constants.ts:21]: and now you can have:  
image

**Review by ryantxu — APPROVED** [review]

**Comment by ryantxu** [comment]
See also https://github.com/grafana/grafana/issues/14554

File: public/app/features/dashboard/state/actions.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dprokop) [reply public/app/features/dashboard/state/actions.ts]: updated :)

**Comment by dragonfly-net** [comment]
You broke singlestat and autoconvert don't work. We cannot migrate from old 6 grafana to 7 now. Please, fix it or return singlestat.

**Comment by dprokop** [comment]
> You broke singlestat and autoconvert don't work. We cannot migrate from old 6 grafana to 7 now. Please, fix it or return singlestat.

@dragonfly-net This is not merged and does not affect Grafana 7. It's to be removed in Grafana 8. If you are struggling updating 6->7 please open an issue.

**Pull request merged**